README.md
[![bitHound Overall Score](https://www.bithound.io/github/flipace/lovli.js/badges/score.svg)](https://www.bithound.io/github/flipace/lovli.js)
[![bitHound Dependencies](https://www.bithound.io/github/flipace/lovli.js/badges/dependencies.svg)](https://www.bithound.io/github/flipace/lovli.js/master/dependencies/npm)
[![bitHound Code](https://www.bithound.io/github/flipace/lovli.js/badges/code.svg)](https://www.bithound.io/github/flipace/lovli.js)

![lovli.js](static/images/logo-small.png)
# lovli.js
A starter kit which allows you to develop realtime applications in minutes.

![lovli.js Preview](static/images/preview.gif)

**Demo:** [lovli.neschkudla.at](http://lovli.neschkudla.at/)

### Stack

- [facebook/react](https://github.com/facebook/react) - View
- [rethinkdb/horizon](https://github.com/rethinkdb/horizon) - Realtime Database Sync
- [flipace/horizon-react](https://github.com/flipace/horizon-react) - Connect View + Data
- [gaearon/react-hot-loader](https://github.com/gaearon/react-hot-loader) - Hot Reloading of React Components
- [webpack/webpack](https://github.com/webpack/webpack) - Builds & Dev-Server

### Installation
``` bash
$ git clone --depth=1 https://github.com/flipace/lovli.js
$ cd lovli.js
$ npm i
```

You'll need to have RethinkDB running.

### Run
``` bash
$ npm start # starts app in dev mode
$ npm run prod # starts server in production mode
$ npm run build # builds source files in .build/
$ node .build/server.bundle.js # starts server (after you built with npm run build)
```

This will start a server listening on port ```3000```.
*You can change the port in* ```config/page.js``` *or by setting the ```PORT``` environment variable*.

### Configure horizon
If you want to configure the props for the horizon client, you can do that
by passing a configuration object as the first parameter to ```HorizonConnector()``` in ```source/client/containers/Root.js``` and the target element as the second parameter.

The props you provide there will be passed to ```horizon()```.

### FAQ

#### How to add vendor scripts and css?
If you want to add any vendor javascript or css, you'll likely not want it to be processed by
babel or localized by css loader.

Add these scripts to the static/vendor folder and they will use a different loader configuration.

#### How to configure data subscriptions?
Check out [flipace/horizon-react](https://github.com/flipace/horizon-react) for documentation of the
horizon-react ```subscribe``` function.

#### Why won't the server reload?
The server won't hot reload. It would be possible to implement a restart on file change though.

### Contributing
Pull Requests are very welcome!

If you find any issues, please report them via [Github Issues](https://github.com/flipace/lovli.js/issues)!

### License
(MIT)


CHANGELOG.md
# Change Log

### v0.9.0
- remove redux dependencies
- allow passing port via env PORT
- update to horizon 2 and change automatic inserts/deletions accordingly (fixes #19)
- 

### v0.8.0
- fixes bug which caused db name from db config not being used
- updated horizon-react to 0.3.2
- updated react-router from [2.4.0 to 2.7.0](https://github.com/reactjs/react-router/blob/master/CHANGES.md)

### v0.7.6
- fix $hz_v$ issue by faking field on server
- set social icons to fixed position

### v0.7.5
- fixes issue with HMR (#11)

### v0.7.4
- updated to horizon client&server to 1.0.3
- removed rxjs dep

### v0.7.3
- temporarily add rxjs@5.0.0-beta.7 as a dependency for lovli.js until new @horizon/client version is out with fixed rxjs dependency
- remove hoist-non-react-statics from deps

### v0.7.2
- added autofocus for input field

### v0.7.1
- stupid me forgot to bump horizon-react version

### v0.7.0
- added import/exports loader until pr merged
- updated to *horizon@1.0.1* #celebrate! https://horizon.io
- updated *horizon-react* to v0.3.0
- updated *react-hot-loader* to 3.0.0-beta2
- creating the horizon client manually in lovli.js instead of letting it be generated by horizon-react
- renamed db from 'horizon' to 'lovli'
- fixed webpack issue with later.js and removed LATER_COV= env
- fixed react-hot-loader HMR issue

### v0.6.0
- updated to ```horizon-react@0.2.1```

### v0.5.0
- added sourcemaps for css & using extract-text-webpack-plugin to extract css to styles.css
- added font extensions to file-loader
- added font-awesome in static/vendor folder to demonstrate adding vendor js/css
- added possibility to add vendor css/js in static/vendor folder (import 'static/...')
- fixed width of "github stars" iframe (2 digits!)
- some README.md additions

### v0.4.1
- better TodoItem styles
- automatically adding random todo ever 2 minutes
- removing jwt token on every load (client)
- added note about todo deletion
- updated @horizon/client to 0.1.0

### v0.4.0
- some readme additions
- added logo and various styles for the example page
- added ```postcss-nested, file-loader, later, rethinkdbdash```
- automatically removing todos every 10 minutes
- nicer todo entry by using input field instead of prompt

### v0.3.1
- updated dependencies to their latest versions

### v0.3.0
- added production configuration
  - run ```npm run prod``` to start the app in production mode
  - run ```npm run build``` to generate the combined js files
  - run ```node .build/server.bundle.js``` to start your app manually
- added page configuration in ```config/page.js```
  - there you can change title, port and horizon token_secret
- fixed webpack config to only include "source" folder for babel-loader

### v0.2.2
- explicitly setting token_secret for horizon server to make sure tokens are valid on server restart
- update @horizon/client to 0.0.4-4

### v0.2.1
- allow env PORT to set port for express
- changes webpack dev server port to 9095
- updates horizon-react to 0.1.2
- adds horizon-react mapDataObject example

### v0.2.0
- moved utils/horizon to horizon-react
- splitted TodoList and AddTodoButton into separate components
- added new subscribe example for horizon-react to TodoList

### v0.1.0
- initial commit


.build/server.bundle.js
/******/ (function(modules) { // webpackBootstrap
/******/ 	// The module cache
/******/ 	var installedModules = {};

/******/ 	// The require function
/******/ 	function __webpack_require__(moduleId) {

/******/ 		// Check if module is in cache
/******/ 		if(installedModules[moduleId])
/******/ 			return installedModules[moduleId].exports;

/******/ 		// Create a new module (and put it into the cache)
/******/ 		var module = installedModules[moduleId] = {
/******/ 			exports: {},
/******/ 			id: moduleId,
/******/ 			loaded: false
/******/ 		};

/******/ 		// Execute the module function
/******/ 		modules[moduleId].call(module.exports, module, module.exports, __webpack_require__);

/******/ 		// Flag the module as loaded
/******/ 		module.loaded = true;

/******/ 		// Return the exports of the module
/******/ 		return module.exports;
/******/ 	}


/******/ 	// expose the modules object (__webpack_modules__)
/******/ 	__webpack_require__.m = modules;

/******/ 	// expose the module cache
/******/ 	__webpack_require__.c = installedModules;

/******/ 	// __webpack_public_path__
/******/ 	__webpack_require__.p = "/assets/";

/******/ 	// Load entry module and return exports
/******/ 	return __webpack_require__(0);
/******/ })
/************************************************************************/
/******/ ([
/* 0 */
/***/ function(module, exports, __webpack_require__) {

	'use strict';

	var _app = __webpack_require__(3);

	var _app2 = _interopRequireDefault(_app);

	function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

	_app2.default.run();

	(function () {
	  function tagSource(fn, localName) {
	    if (typeof fn !== "function") {
	      return;
	    }

	    if (fn.hasOwnProperty("__source")) {
	      return;
	    }

	    try {
	      Object.defineProperty(fn, "__source", {
	        enumerable: false,
	        configurable: true,
	        value: {
	          fileName: '/Users/patrickneschkudla/Development/open-source/lovli.js/source/server/entry.server.js',
	          localName: localName
	        }
	      });
	    } catch (err) {}
	  }
	})();

/***/ },
/* 1 */
/***/ function(module, exports) {

	module.exports = {
	  title: 'lovli.js | Hello World',
	  port: 3000,
	  token_secret: 'thisislovlijs'
	};


/***/ },
/* 2 */
/***/ function(module, exports) {

	const host = '127.0.0.1';
	const webpackPort = 9095;

	module.exports = {
	  host,
	  webpackPort,
	  baseUrl: `http://${host}:${webpackPort}`
	};


/***/ },
/* 3 */
/***/ function(module, exports, __webpack_require__) {

	'use strict';

	Object.defineProperty(exports, "__esModule", {
	  value: true
	});

	var _path = __webpack_require__(6);

	var _path2 = _interopRequireDefault(_path);

	var _express = __webpack_require__(5);

	var _express2 = _interopRequireDefault(_express);

	var _server = __webpack_require__(4);

	var _server2 = _interopRequireDefault(_server);

	var _devProps = __webpack_require__(2);

	var _devProps2 = _interopRequireDefault(_devProps);

	var _page = __webpack_require__(1);

	var _page2 = _interopRequireDefault(_page);

	function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

	var app = (0, _express2.default)();

	app.use('/static', _express2.default.static(_path2.default.join(process.cwd(), '.build')));

	var bundle =  true ? '/static/client.bundle.js' : 'http://127.0.0.1:' + _devProps2.default.webpackPort + '/static/client.bundle.js';

	app.use('/', function (req, res) {
	  res.status(200).send('<!doctype html>\n    <html>\n      <head>\n        <title>' + _page2.default.title + '</title>\n      </head>\n      <body>\n        <div id=\'root\'></div>\n        <script src="' + bundle + '"></script>\n      </body>\n    </html>');
	});

	var run = function run() {
	  var port = process.env.PORT || _page2.default.port;

	  var httpServer = app.listen(port, function (err) {
	    if (err) {
	      console.log(err); // eslint-disable-line
	      return;
	    }

	    console.log('Express listening at http://localhost:' + port); // eslint-disable-line
	  });

	  // if we want to start secure, add key and cert props
	  var horizonServer = (0, _server2.default)(httpServer, {
	    auto_create_table: true,
	    auto_create_index: true,
	    auth: {
	      allow_anonymous: true,
	      allow_unauthenticated: true,
	      token_secret: _page2.default.token_secret
	    }
	  });
	};

	var _default = {
	  run: run
	};
	exports.default = _default;

	(function () {
	  function tagSource(fn, localName) {
	    if (typeof fn !== "function") {
	      return;
	    }

	    if (fn.hasOwnProperty("__source")) {
	      return;
	    }

	    try {
	      Object.defineProperty(fn, "__source", {
	        enumerable: false,
	        configurable: true,
	        value: {
	          fileName: '/Users/patrickneschkudla/Development/open-source/lovli.js/source/server/app.js',
	          localName: localName
	        }
	      });
	    } catch (err) {}
	  }

	  tagSource(app, 'app');
	  tagSource(bundle, 'bundle');
	  tagSource(run, 'run');
	  tagSource(_default, 'default');
	})();

/***/ },
/* 4 */
/***/ function(module, exports) {

	module.exports = require("@horizon/server");

/***/ },
/* 5 */
/***/ function(module, exports) {

	module.exports = require("express");

/***/ },
/* 6 */
/***/ function(module, exports) {

	module.exports = require("path");

/***/ }
/******/ ]);